src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using System.Linq;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private readonly MethodSymbol _method;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly SyntheticBoundNodeFactory _fact;
+        private NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _diagnostics = diagnostics;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            if (!method.Parameters.Any(x => x is SourceParameterSymbolBase param
+                                            && param.IsNullChecked))
+            {
+                return body;
+            }
+
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return (BoundStatement)rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            return AddNullChecksToBody(node);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body)
+        {
+            var statementList = ArrayBuilder<BoundStatement>.GetInstance();
+            foreach (ParameterSymbol x in _method.Parameters)
+            {
+                if (x is SourceParameterSymbolBase param
+                    && param.IsNullChecked)
+                {
+                    Debug.Assert(!param.Type.IsValueType || param.Type.SpecialType == SpecialType.System_Nullable_T);
```

Comment:
>param.Type.SpecialType [](start = 60, length = 22)

This will fail for constructed types such as `int?`. There is an `IsNullableType()` extension method that can be used instead.